mlir/lib/Dialect/MIOpen/utility/transformMapUtils.cpp

```diff
@@ -252,34 +297,39 @@ using VectorizationData = llvm::IndexedMap<Optional<int64_t>>;
 /// unit stride (the dimension that broke things could have jumps, padding,
 /// etc.).
 template <typename T>
-static Optional<int64_t>
+static Optional<VectorizationInfo>
 propagateUnmergeVectorization(T &&dimAndLength,
                               const VectorizationData &input) {
+  Optional<VectorizationInfo> result;
   int64_t previousDimsStride = 1;
-  Optional<int64_t> maxVectorizationLength;
   for (auto pair : dimAndLength) {
     uint32_t upperDim = std::get<0>(pair);
     int64_t dimLength = std::get<1>(pair);
     if (input[upperDim].hasValue()) {
-      if (!maxVectorizationLength.hasValue())
-        maxVectorizationLength = 1;
-      if (maxVectorizationLength == previousDimsStride) {
-        maxVectorizationLength =
-            (*maxVectorizationLength) * input[upperDim].getValue();
+      VectorizationInfo upperInfo = *input[upperDim];
```

Comment:
We won't hit that assert because of the call to `grow()` at the top.

And I'm using `IndexedMap` to be able to have a nice interface around `SmellVector<Option<VectorizationInfo>>` so that I don't have to manually write out the grow() method - why not use the type that's already there?

I'd be willing to go in for some other small dense map primitive, but I figure the space costs of summoning up some `None`s are much less than the bother of having to go through hash functions.

`IndexedMap` is a map from the dimension to the two integers we're storing as `VectorizationInfo`

Reply:
I didn't comprehend the index conversion from pair of integer to the smallVector index. How does that work? If there is a pair of (5, 10), does that get converted to 15 or 50? 
 - I think it does matter if we allocate 50 elements for 1 value. 
 - On the other hand, if it get converted to 15, then (10, 5) also get converted to 15 so how does it handle collision?

Reply:
... that's not what IndexedMap is

Reply:
`IndexedMap<T>` == `SmallVector<T>` with the constraint that `T` has a "default" value.

That is, in this case `input[upperDim]` gives you an `Option<VectorizationInfo>` from the underlying array.

The pairs of integers are the _values_, not the keys

Reply:
Ahhh that's right, I did see `IndexedMap` using `SmallVector` under the hood, but was confused by what that `identity<T>` does. I was assuming it is doing vector version of hashmap (maybe because of there's a "map" in its naming?) and that identity function is doing something smart until today when I poked I confirm that index is really just a integer. Sorry for the confusion.